File: dump-dude/JS-Base/Pure-Base/Queue.js
/**Queue : FIFO
 * List of operation performed by queue
        enqueue(): Adds an item at the tail of the queue.
        dequeue(): Removes an item from the head of the queue.
        front(): Retruns the first item in the queue.
        rear(): Retruns the last item in the queue.
        size(): Returns the size of the queue.
        isEmpty(): Returns true if queue is empty, false otherwise.
        
 */
var Queue = function (size) {
    this.elements = [size];
    this.front = 0;
    this.rear = -1;
}
Queue.prototype.EnQueue = function (val) {
    this.rear++;
    this.elements[this.rear] = val;
}
Queue.prototype.Peek = function () {
    return this.elements[this.front];
}
Queue.prototype.DeQueue = function () {
    this.elements.shift();
}
Queue.prototype.isEmpty = function () {
    if (this.rear === this.elements.length)
        return true;
    else
        return false;
}
Queue.prototype.isFull = function () {
    if (this.rear + 1 === this.elements.length)
        return true;
    else
        return false;
}
Queue.prototype.getFrontItem = function () {
    return this.elements[this.front];
}
Queue.prototype.getRearItem = function () {
    return this.elements[this.rear];
}

var myQueue = new Queue(4);
myQueue.EnQueue(1);
myQueue.EnQueue(2);
myQueue.EnQueue(3);
myQueue.EnQueue(4);
console.log(`isFull : ${myQueue.isFull()}`);
console.log(` getFrontItem ${myQueue.getFrontItem()}`);
console.log(` Peek ${myQueue.Peek()}`);
console.log(`isEmpty : ${myQueue.isEmpty()}`);
myQueue.DeQueue();
myQueue.DeQueue();
console.log(` Peek ${myQueue.Peek()}`);
myQueue.DeQueue();
myQueue.DeQueue();

File: dump-dude/CS-Base/Queue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace next_gen
{
    class Queue
    {
        /*
        * FIFO = First In First Out
        * Insertion allow at one end - Front
        * Deletion allow at one end - Rear
        */
        private int[] queueArray;
        private int front;
        private int rear;
        public Queue()
        {
            queueArray = new int[10];
            front = -1;
            rear = -1;
        }
        public Queue(int maxSize)
        {
            queueArray = new int[maxSize];
            front = -1;
            rear = -1;
        }
        public bool isEmpty()
        {
            return (front == -1 || front == rear - 1);
        }
        public int size()
        {
            return rear + 1;
        }
        public bool isFull()
        {
            return (rear == queueArray.Length - 1);
        }
        public void queue(int item)
        {
            if (isFull())
            {
                Console.WriteLine("Queue Over flow!");
                return;
            }
            front = 0;
            rear++;
            queueArray[rear] = item;
        }
        public int deQueue()
        {
            int x;
            if (isEmpty())
            {
                Console.WriteLine("Queue is empty!");
                return -1;
            }
            if (front == 0)
            {
                front--;
            }
            x = queueArray[rear];
            rear--;
            return x;

        }
        public int peek()
        {
            if (isEmpty())
            {
                Console.WriteLine("Queue is empty");
                return -1;
            }
            return queueArray[rear];
        }
    }
}
